sklearn/pipeline.py

```diff
@@ -177,19 +177,25 @@ def _validate_steps(self):
         for t in transformers:
             if t is None or t == 'passthrough':
                 continue
-            if (not (hasattr(t, "fit") or hasattr(t, "fit_transform")) or not
-                    hasattr(t, "transform")):
+            if (not hasattr(t, "fit_transform") or
+                    (not hasattr(t, "fit") and hasattr(t, "transform"))):
                 raise TypeError("All intermediate steps should be "
-                                "transformers and implement fit and transform "
-                                "or be the string 'passthrough' "
-                                "'%s' (type %s) doesn't" % (t, type(t)))
+                                "transformers and implement fit and "
+                                "transform, fit_transform or be the string "
+                                "'passthrough'. '%s' (type %s) doesn't"
+                                % (t, type(t)))
+            elif not hasattr(t, "transform"):
```

Comment:
Don't you want to also raise a warning if the last step doesn't have a `transform` or `predict` method? The pipeline is also not reusable in those cases.

Reply:
Right, I added a a check for the final estimator as well.